File: resources/python_old/vector.py
from settings import *
import math

class  Vector:
    def __init__(self, x, y):
        self._x = x
        self._y = y
    
    def setX(self, x):
        self._y = x

    def setY(self, y):
        self._y = y
    
    def getX(self):
        return self._x

    def getY(self):
        return self._y

    def setAngle(self, angle):
        length = self.length()
        self._x = math.cos(angle) * length
        self._y = math.sin(angle) * length
    
    def angle(self):
        return math.atan2(self._y, self._x)
    
    def setLength(self, length):
        angle = self.angle()
        self._x = math.cos(angle) * length
        self._y = math.sin(angle) * length
    
    def length(self):
        return math.sqrt( ( self._x * self._x ) + ( self._y * self._y ) )

    def distance(self, vector): 
        dx = vector.getX() - self._x;
        dy = vector.getY() - self._y;

        return math.sqrt(dx * dx + dy * dy);

    def add(self, vector):
        return Vector(self._x + vector.getX(), self._y + vector.getY());

    def subtract(self, vector):
        return Vector(self._x - vector.getX(), self._y - vector.getY());

    def multiply(self, vector):
        return Vector(self._x * vector.getX(), self._y * vector.getY());

    def divide(self, vector):
        return Vector(self._x / vector.getX(), self._y / vector.getY());

    def addTo(self, vector):
        self._x += vector.getX()
        self._y += vector.getY()

    def subtractFrom(self, vector):
        self._x -= vector.getX()
        self._y -= vector.getY()

    def multiplyBy(self, value):
        self._x *= value
        self._y *= value

    def divideBy(self, value):
        self._x /= value
        self._y /= value

    # def scale(vector):
    #     return Vector(self._x * vector, this.y * vector);

# export class Vector {
#     constructor(x, y) {
#         this.x = x;
#         this.y = y;
#     }

#     length() {
#         return Math.sqrt(this.x * this.x + this.y * this.y);
#     }

#     setLength(value) {
#         const ratio = value / this.length;
#         this.x *= ratio;
#         this.y *= ratio;
#     }

#     add(vector) {
#         return new Vector(this.x + vector.x, this.y + vector.y);
#     }

#     mult(value) {
#         return new Vector(this.x * value, this.y * value);
#     }

#     sub(vector) {
#         return new Vector(this.x - vector.x, this.y - vector.y);
#     }

#     distance(vector) {
#         const dx = vector.x - this.x;
#         const dy = vector.y - this.y;

#         return Math.sqrt(dx * dx + dy * dy);
#     }



#     scale(vector) {
#         return new Vector(this.x * vector, this.y * vector);
#     }

#     magnitude() {
#         return Math.sqrt(this.x * this.x + this.y * this.y);
#     }

#     normalize() {
#         const length = this.length();
#         if (length == 0)
#             return this;

#         return new Vector(this.x / length, this.y / length);
#     }

#     unit() {
#         const magnitude = this.magnitude();
#         return new Vector(this.x / magnitude, this.y / magnitude);
#     }

#     dot(vector) {
#         return this.x * vector.x + this.y * vector.y;
#     }

#     cross(vector) {
#         return this.x * vector.y - this.y * vector.x;
#     }

#     rotate(angle) {
#         const cos = Math.cos(angle);
#         const sin = Math.sin(angle);
#         return new Vector(this.x * cos - this.y * sin, this.x * sin + this.y * cos);
#     }

#     angle(vector2) {
#         return Math.atan2(this.cross(vector2), this.dot(vector2));
#     }

#     project(vector2) {
#         const scale = this.dot(vector2) / vector2.dot(vector2);
#         return vector2.scale(scale);
#     }

#     getOverlap(otherVector) {
#         return Math.min(this.x, otherVector.x) * Math.min(this.y, otherVector.y);
#     }

#     perpendicular() {
#         return new Vector(-this.y, this.x);
#     }

#     projectOnto(vector) {
#         const unit = this.unit();
#         return unit.mult(unit.dot(vector));
#     }
# }

File: js/iot-engine/helpers/Vector.js
class Vector {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    static zero() {
        return new Vector(0, 0);
    }

    static $(x = 0, y = 0) {
        return new Vector(x, y);
    }

    isEmpty() {
        return (this.getX() == 0 && this.getY() == 0);
    }

    isNull() {
        return (this.getX() == null || this.getY() == null);
    }

    normalize() {
        const length = this.getLength();
        if (length == 0)
            return this;

        return new Vector(this.x / length, this.y / length);
    }

    dot(v) {
        return this.x * v.x + this.y * v.y;
    }

    cross(v) {
        return this.x * v.y - this.y * v.x;
    }

    rotate(angle) {
        let cosA = Math.cos(angle);
        let sinA = Math.sin(angle);
        return new Vector(this.x * cosA - this.y * sinA, this.x * sinA + this.y * cosA);
    }

    setX(value) {
        this.x = value;
    }

    getX() {
        return this.x;
    }

    setY(value) {
        this.y = value;
    }

    getY() {
        return this.y;
    }

    setAngle(angle) {
        var length = this.getLength();
        this.x = Math.cos(angle) * 1;
        this.y = Math.sin(angle) * 1;
    }

    getAngle() {
        return Math.atan2(this.y, this.x);
    }

    heading() {
        var angle = this.getAngle();
        return new Vector(Math.cos(angle), Math.sin(angle));
    }

    getHeading(other) {
        return new Vector(Math.cos(other), Math.sin(other));
    }

    setHeading(angle) {
        var angle = this.getAngle();
        this.x = Math.cos(angle);
        this.y = Math.sin(angle);
    }

    setLength(length) {
        var angle = this.getAngle();
        this.x = Math.cos(angle) * length;
        this.y = Math.sin(angle) * length;
    }

    getLength() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    length() {
        return this.getLength();
    }

    subtract(v) {
        return new Vector(this.x - v.x, this.y - v.y);
    }

    add(v) {
        return new Vector(this.x + v.x, this.y + v.y);
    }

    multiply(val) {
        return new Vector(this.x * val, this.y * val);
    }

    divide(scalar) {
        return new Vector(this.x / scalar, this.y / scalar);
    }

    addTo(v2) {
        this.x += v2.getX();
        this.y += v2.getY();

        return this;
    }

    subtractFrom(v2) {
        this.x -= v2.getX();
        this.y -= v2.getY();

        return this;
    }

    multiplyBy(scalar) {
        return new Vector(this.x * scalar, this.y * scalar);
    }

    multiplyByVector(v) {
        var x = this.x * v.x;
        var y = this.y * v.y;
        
        return new Vector(x, y);
    }

    divideBy(val) {
        this.x /= val;
        this.y /= val;

        return this;
    }

    addVal(v) {
        return new Vector(this.x + v, this.y + v);
    }

    addValBy(v) {
        this.x += v;
        this.y += v;
        return this;
    }

    clone() {
        return new Vector(this.x, this.y);
    }

    // distance(p1) {
    //     var dx = this.x - p1.x,
    //         dy = this.y - p1.y;

    //     return Math.sqrt(dx * dx + dy * dy);
    // }

    distance(p1) {
        if (!p1 || !(p1 instanceof Vector)) {
            console.error("Invalid parameter passed to distance method.");
            return 0;
        }
        var dx = this.x - p1.x,
            dy = this.y - p1.y;
    
        return Math.sqrt(dx * dx + dy * dy);
    }
    
    angle(other) {
        return Math.atan2(other.y - this.y, other.x - this.x);
    }


    negate() {
        this.x = -this.x;
        this.y = -this.y;

        return this;
    }

    unit(){
        if(this.mag() === 0){
            return new Vector(0,0);
        } else {
            return new Vector(this.x/this.mag(), this.y/this.mag());
        }
    }

    mag(){
        return Math.sqrt(this.x**2 + this.y**2);
    }

    mult(n){
        return new Vector(this.x*n, this.y*n);
    }

    normalize() {
        let length = this.getLength();
        return new Vector(this.x / length, this.y / length);
    }

    drawVec(start_x, start_y, n, color){
        ctx.beginPath();
        ctx.moveTo(start_x, start_y);
        ctx.lineTo(start_x + this.x * n, start_y + this.y * n);
        ctx.strokeStyle = color;
        ctx.stroke();
        ctx.closePath();
    }
}